semaphore/semaphore_test.go

```diff
@@ -0,0 +1,91 @@
+package semaphore_test
+
+import (
+	"fmt"
+	"time"
+	"testing"
+	"context"
+	
+	"github.com/madlambda/spells/assert"
+	"github.com/madlambda/spells/semaphore"
+)
+
+func TestSemaphore(t *testing.T) {
+	sizes := []uint{ 1, 2, 3, 1000 }
+	
+	for _, size := range sizes {
+		t.Run(fmt.Sprintf("Size%d", size), func(t *testing.T) {
+			s := semaphore.New(size)
+			// WHY: each test function releases all acquires
+			// we use this to test if the semaphore will reset to
+			// its initial state properly after all releases.
+			testSemaphore(t, s, size)
+			testSemaphore(t, s, size)
+		})
+	}
+	
+}
+
+func TestSemaphoreSizeCantBeZero(t *testing.T) {
+	defer func() {
+        if r := recover(); r == nil {
```

Comment:
maybe ignore r.(runtime.Error) otherwise test would pass for runtime panics also.